rust/template/differential_datalog/src/program/worker.rs

```diff
@@ -428,10 +442,34 @@ impl<'a> DDlogWorker<'a> {
                     HashMap::with_capacity_and_hasher(program.nodes.len(), FnvBuildHasher::default());
             let mut arrangements = FnvHashMap::default();
 
+            // Create an `Enabled` relation used to enforce the dataflow termination in the
+            // presence of delayed relations.  A delayed relation can potentially generate an
+            // infinite sequence of outputs for all future timestamps, preventing the
+            // differential dataflow from terminating.  DD can only terminate cleanly when there
+            // is no data in flight, and will keep spinning forever if new records keep getting
+            // injected in the dataflow.  To prevent this scenario, we join all delayed relations
+            // with `Enabled`.  `Enabled` contains a single record (an empty tuple) as long as
+            // the program is running.  We retract this record on shutdown, hopefully enforcing
+            // quiescing the dataflow.
+            let (enabled_session, enabled_collection) = outer.new_collection::<(),Weight>();
```

Comment:
I'd add a note pointing to [timely-dataflow/#306](https://github.com/TimelyDataflow/timely-dataflow/issues/306) as the (hopefully) ultimate solution to this

Reply:
Ack, although it doesn't look like it will be implemented any time soon...